

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Getting Started Tutorial &mdash; k-seq 1.0.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="k-seq package" href="k-seq_package.html" />
    <link rel="prev" title="Installation" href="installation.html" />
    <link href="_static/update_theme.css" rel="stylesheet" type="text/css">

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> k-seq
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Getting Started Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data-preprocessing">Data preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#join-paired-end-reads-and-count-unique-sequences-using-easydiver">Join paired-end reads and count unique sequences using <code class="docutils literal notranslate"><span class="pre">EasyDIVER</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#load-count-files">Load count files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filter-sequences-and-samples">Filter sequences and samples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sequence-quantification">Sequence quantification</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#quantify-the-sequence">Quantify the sequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="#calculate-the-dependent-variable">Calculate the dependent variable</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#kinetic-model-fitting">Kinetic model fitting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#least-squares-fitting-with-bootstrapping">Least-squares fitting with bootstrapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fitting-results">Fitting results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="k-seq_package.html"><code class="docutils literal notranslate"><span class="pre">k-seq</span></code> package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">k-seq</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Getting Started Tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/getting_started.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="getting-started-tutorial">
<h1>Getting Started Tutorial<a class="headerlink" href="#getting-started-tutorial" title="Permalink to this headline">¶</a></h1>
<p>This brief tutorial walks through a basic pipeline for <em>k</em>-Seq experiment data analysis using <code class="docutils literal notranslate"><span class="pre">k-seq</span></code> package. The
pipeline includes following steps:</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="#data-preprocessing">Data preprocessing</a></p></li>
<li><p><a class="reference internal" href="#sequence-quantification">Sequence quantification</a></p></li>
<li><p><a class="reference internal" href="#kinetic-model-fitting">Kinetic model fitting</a></p></li>
</ol>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="installation.html"><span class="doc">Installation</span></a> for setting up the environment and installing <cite>k-seq</cite> package.
<a class="reference internal" href="k-seq_package.html"><span class="doc">k-seq package</span></a> for the complete API documentation of the package.</p>
</div>
<div class="section" id="data-preprocessing">
<h2>Data preprocessing<a class="headerlink" href="#data-preprocessing" title="Permalink to this headline">¶</a></h2>
<p>Sequencing data (FASTQ files) are first processed in the following steps:</p>
<ol class="arabic simple">
<li><p>trim/join paired-end reads and deduplicate joined reads as sequence counts</p></li>
<li><p>compose multiple count files as a count table</p></li>
<li><p>filter sequences or samples for downstream analysis</p></li>
</ol>
<div class="section" id="join-paired-end-reads-and-count-unique-sequences-using-easydiver">
<h3>Join paired-end reads and count unique sequences using <code class="docutils literal notranslate"><span class="pre">EasyDIVER</span></code><a class="headerlink" href="#join-paired-end-reads-and-count-unique-sequences-using-easydiver" title="Permalink to this headline">¶</a></h3>
<p>We integrated <a class="reference external" href="https://github.com/ichen-lab-ucsb/EasyDIVER">EasyDIVER</a> pipeline in the script <code class="docutils literal notranslate"><span class="pre">fastq-to-counts.sh</span></code>
where paired-end reads are joined using <a class="reference external" href="https://github.com/neufeld/pandaseq">PANDASeq</a>, reads from sample samples
are pooled, and joined reads are further deduplicated into counts (number of reads) for each unique sequence.</p>
<p><code class="docutils literal notranslate"><span class="pre">fastq-to-counts.sh</span></code> requires all FASTQ files are organized in a folder with file names formatted as:
<code class="docutils literal notranslate"><span class="pre">[sample-name]_L[lane</span> <span class="pre">number]_R[1/2]_001.fastq.gz</span></code>. <code class="docutils literal notranslate"><span class="pre">sample-name</span></code> is the string identifier for an experimental
sample and <code class="docutils literal notranslate"><span class="pre">lane</span> <span class="pre">number</span></code> indicate the lane of the flow-cell. <code class="docutils literal notranslate"><span class="pre">R[1/2]</span></code> indicate the direction of paired-end reads
(<code class="docutils literal notranslate"><span class="pre">R1</span></code> for forward reads and <code class="docutils literal notranslate"><span class="pre">R2</span></code> for reverse reads) that each sample-lane expects to have matched files in both
directions.</p>
<p>Here is an example input file folder contains paired-end reads from two samples <code class="docutils literal notranslate"><span class="pre">R4A-0A_S7</span></code> and <code class="docutils literal notranslate"><span class="pre">R4B-0A_S21</span></code>
with 4 lanes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L001_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L001_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L001_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L001_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L002_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L002_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L002_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L002_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L003_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L003_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L003_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L003_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L004_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L004_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L004_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L004_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>To run <code class="docutils literal notranslate"><span class="pre">EasyDIVER</span></code> pipeline:</p>
<div class="highlight-shell notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10</pre></div></td><td class="code"><div class="highlight"><pre><span></span>&gt; fastq-to-counts.sh <span class="se">\</span>
      -i /path/to/input/folder <span class="se">\</span>
      -o /path/to/output/folder <span class="se">\</span>
      -p CTACGAATTC <span class="se">\ </span>             <span class="c1"># forward adapter sequence for trimming</span>
      -q CTGCAGTGAA <span class="se">\ </span>             <span class="c1"># reverse adapter sequence for trimming</span>
      -c <span class="se">\ </span>                        <span class="c1"># use completely matching in joining and discard reads with error</span>
      -a <span class="se">\ </span>                        <span class="c1"># join the paired-end reads first before trimming the adapters for reads with heavily overlapped regions</span>
      -T <span class="m">12</span>                        <span class="c1"># number of threads to run in parallel for pandaSeq</span>
&gt; ls /path/to/output/folder
counts  fastas  fastqs  histos  log.txt
</pre></div>
</td></tr></table></div>
<p>The output folder contains following files:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">counts</span></code>: deduplicated count files named as <code class="docutils literal notranslate"><span class="pre">[sample-name]_counts.txt</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fastqs</span></code>: joined FASTQ files named as <code class="docutils literal notranslate"><span class="pre">[sample-name].joined.fastq</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fastas</span></code>: joined FASTA files named as <code class="docutils literal notranslate"><span class="pre">[sample-name].joined.fasta</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">histos</span></code>: files with the joined reads length histogram statistics, named as <code class="docutils literal notranslate"><span class="pre">[sample-files]_counts_histo.txt</span></code></p></li>
</ul>
<p>We will use the count files for each sample under the <code class="docutils literal notranslate"><span class="pre">counts</span></code> folder, which has following format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">number</span> <span class="n">of</span> <span class="n">unique</span> <span class="n">sequences</span> <span class="o">=</span>     <span class="mi">981824</span>
<span class="n">total</span> <span class="n">number</span> <span class="n">of</span> <span class="n">molecules</span> <span class="o">=</span>    <span class="mi">10096876</span>

<span class="n">GGGGGGGGATTCATGACTATT</span>                                                                 <span class="mi">1</span>
<span class="n">GGGGGGGAGTAGGACTGCAAA</span>                                                                 <span class="mi">1</span>
<span class="n">GGGGGGGAAGACTCCGGAACG</span>                                                                 <span class="mi">1</span>
<span class="n">GGGGGGGAACGCATTTCACGG</span>                                                                 <span class="mi">1</span>
<span class="n">GGGGGGGACGTTCACCGGCAA</span>                                                                 <span class="mi">1</span>
<span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="load-count-files">
<h3>Load count files<a class="headerlink" href="#load-count-files" title="Permalink to this headline">¶</a></h3>
<p>We next use <code class="docutils literal notranslate"><span class="pre">k-seq</span></code> data class <a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData" title="k_seq.data.seq_data.SeqData"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.SeqData</span></code></a> to load and pool count files
into a Python object with merged count table for further analysis.</p>
<p>In this example, we use <a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData.from_count_files" title="k_seq.data.seq_data.SeqData.from_count_files"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SeqData.from_count_files</span></code></a> to load all
count files provided in the <code class="docutils literal notranslate"><span class="pre">k-seq</span></code> repository
(<a class="reference external" href="https://github.com/ichen-lab-ucsb/k-seq/tree/master/data/counts">k-seq/data/counts</a> folder).
We filter files in the folder to load only count files whose file names contain <code class="docutils literal notranslate"><span class="pre">test-d</span></code> and two
DNA quantification data files (<code class="docutils literal notranslate"><span class="pre">total-rna.csv</span></code> and <code class="docutils literal notranslate"><span class="pre">spike-in-rna.csv</span></code>) should not be included.
The independent variable (x, initial BYO concentration) is provided in the file <code class="docutils literal notranslate"><span class="pre">sample-substrate-concentration.csv</span></code>,
which can be passed to the dataset using <code class="docutils literal notranslate"><span class="pre">x_values</span></code> argument.
See <a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData.from_count_files" title="k_seq.data.seq_data.SeqData.from_count_files"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SeqData.from_count_files</span></code></a> for the complete usage.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Files are usually named with relevant information to the sample, such as such as experimental condition,
replicate number, sample name, etc. The loading function
<a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData.from_count_files" title="k_seq.data.seq_data.SeqData.from_count_files"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SeqData.from_count_files</span></code></a> can automatically extract
these metadata from file names using argument <code class="docutils literal notranslate"><span class="pre">name_template</span></code>. The examples count files are named as
<code class="docutils literal notranslate"><span class="pre">test-d-{exp_cond:</span> <span class="pre">A,</span> <span class="pre">B,</span> <span class="pre">C,</span> <span class="pre">...}{rep_id:</span> <span class="pre">1,</span> <span class="pre">2,</span> <span class="pre">3}-_S{sample_id}_counts.txt</span></code> and the samples are named
as <code class="docutils literal notranslate"><span class="pre">{experimental</span> <span class="pre">condition}{rep}</span></code>. Use <code class="docutils literal notranslate"><span class="pre">name_template</span></code> to locate the metadata (using <code class="docutils literal notranslate"><span class="pre">{}</span></code>) and sample
name (using <code class="docutils literal notranslate"><span class="pre">[]</span></code>) in file name with corresponding data type (<code class="docutils literal notranslate"><span class="pre">int</span></code>, <code class="docutils literal notranslate"><span class="pre">float</span></code>, or <code class="docutils literal notranslate"><span class="pre">str</span></code> by default).
See below for example code.</p>
</div>
<p>Example code:</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">k_seq.data</span> <span class="kn">import</span> <span class="n">SeqData</span>

<span class="n">sub_c</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/sample-substrate-concentration.csv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">squeeze</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">dataset</span> <span class="o">=</span> <span class="n">SeqData</span><span class="o">.</span><span class="n">from_count_files</span><span class="p">(</span>
    <span class="n">count_files</span><span class="o">=</span><span class="s1">&#39;path/to/counts&#39;</span><span class="p">,</span>
    <span class="n">pattern_filter</span><span class="o">=</span><span class="s1">&#39;test-d-&#39;</span><span class="p">,</span>
    <span class="n">name_template</span><span class="o">=</span><span class="s1">&#39;test-d-[</span><span class="si">{cond}</span><span class="s1">{exp_rep, int}]_S{sample, int}_counts.txt&#39;</span><span class="p">,</span>
    <span class="n">sort_by</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span>
    <span class="n">x_values</span><span class="o">=</span><span class="n">sub_c</span><span class="p">,</span>
    <span class="n">x_unit</span><span class="o">=</span><span class="s1">&#39;M&#39;</span><span class="p">,</span>    <span class="c1"># substrate concentration unit: M</span>
    <span class="n">input_sample_name</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;R0&#39;</span><span class="p">],</span>
    <span class="n">note</span><span class="o">=</span><span class="s1">&#39;Example k-seq data&#39;</span>
<span class="p">)</span>
</pre></div>
</td></tr></table></div>
<p><code class="docutils literal notranslate"><span class="pre">dataset</span></code> is a <a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData" title="k_seq.data.seq_data.SeqData"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.SeqData</span></code></a> object and contains
a count table (<a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqTable" title="k_seq.data.seq_data.SeqTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.SeqTable</span></code></a>, a subclass of <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>)
at <code class="docutils literal notranslate"><span class="pre">dataset.table.original</span></code>.</p>
</div>
<div class="section" id="filter-sequences-and-samples">
<h3>Filter sequences and samples<a class="headerlink" href="#filter-sequences-and-samples" title="Permalink to this headline">¶</a></h3>
<p>Depending on the experimental design and quality control criteria, sequences or samples sometimes need to be filtered.
<code class="docutils literal notranslate"><span class="pre">k-seq</span></code> provides several filters for this purpose. For example, sequences contain ambiguous nucleotides (‘N’) or
with undesired length might need to be removed, and the spike-in sequence should be excluded from fitting. We filter
these sequences and saved a filtered table under <code class="docutils literal notranslate"><span class="pre">dataset.table</span></code>:</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">k_seq.data.filters</span> <span class="kn">import</span> <span class="n">NoAmbiguityFilter</span><span class="p">,</span> <span class="n">SeqLengthFilter</span><span class="p">,</span> <span class="n">SpikeInFilter</span>

<span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered</span> <span class="o">=</span> <span class="n">SeqLengthFilter</span><span class="p">(</span><span class="n">min_len</span><span class="o">=</span><span class="mi">21</span><span class="p">,</span> <span class="n">max_len</span><span class="o">=</span><span class="mi">21</span><span class="p">)(</span>
    <span class="n">NoAmbiguityFilter</span><span class="p">()(</span>
        <span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">original</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered</span> <span class="o">=</span> <span class="n">SpikeInFilter</span><span class="p">(</span>
    <span class="n">center_seq</span><span class="o">=</span><span class="s1">&#39;AAAAACAAAAACAAAAACAAA&#39;</span><span class="p">,</span>
    <span class="n">radius</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">dist_type</span><span class="o">=</span><span class="s1">&#39;edit&#39;</span><span class="p">,</span>       <span class="c1"># we consider all sequences within 2 edit distance to center_seq as spike-in</span>
    <span class="n">target</span><span class="o">=</span><span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered</span>     <span class="c1"># target table needed to calculate sequence edit distance to center_seq</span>
<span class="p">)(</span><span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
</div>
</div>
<div class="section" id="sequence-quantification">
<h2>Sequence quantification<a class="headerlink" href="#sequence-quantification" title="Permalink to this headline">¶</a></h2>
<p>Once a count table with sequences of interests are prepared, we can further quantify the dependent variable in the
kinetic model (e.g., fraction reacted) for fitting.</p>
<div class="section" id="quantify-the-sequence">
<h3>Quantify the sequence<a class="headerlink" href="#quantify-the-sequence" title="Permalink to this headline">¶</a></h3>
<p>Depending on the design of the <em>k</em>-Seq experiment, different quantification methods could be used. The <code class="docutils literal notranslate"><span class="pre">k-seq</span></code>
package implemented two common quantification approaches: 1) by total sequence amount and 2) by amount of
external standards (i.e., spike-in sequence).</p>
<p>Total sequence (DNA/RNA/protein) amount in each sample can be measured using proper experimental technics
(e.g., qPCR for RNA amount). The quantity of each sequence in a sample can be calculated as the ratio of sequence’s
counts over the total number of reads in the sample, multiplied by the total sequence amount.
In our examples, the total RNA amount for each sample is saved in <code class="docutils literal notranslate"><span class="pre">total-rna.csv</span></code>, and we can pass it to
<code class="docutils literal notranslate"><span class="pre">dataset</span></code> using method <a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData.add_sample_total" title="k_seq.data.seq_data.SeqData.add_sample_total"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SeqData.add_sample_total</span></code></a>. Please note
that in order to calculate the fraction of reads for each sequence in a sample, we need to specify a <code class="docutils literal notranslate"><span class="pre">full_table</span></code>,
which is, in our case, the filtered table to exclude added spike-in sequence (and a small number of sequences discarded).</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="n">total_rna</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;path/to/total-rna.csv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">squeeze</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># add sample total info</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">add_sample_total</span><span class="p">(</span>
    <span class="n">total_amounts</span><span class="o">=</span><span class="n">total_rna</span><span class="p">,</span>
    <span class="n">full_table</span><span class="o">=</span><span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered</span><span class="p">,</span>
    <span class="n">unit</span><span class="o">=</span><span class="s1">&#39;ng&#39;</span>
<span class="p">)</span>
<span class="c1"># quantify fitlered seqeunces</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered_ng</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">sample_total</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<p>Alternatively, we can quantify RNA amount with spike-in sequence using method
<a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData.add_spike_in" title="k_seq.data.seq_data.SeqData.add_spike_in"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SeqData.add_spike_in</span></code></a>. The amount of a sequence is calculated as the
ratio of the sequence’s counts over the counts of spike-in sequences, multiplied by the spike-in amount. In this case,
we choose a <code class="docutils literal notranslate"><span class="pre">base_table</span></code> as the original table with spike-read reads.</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="n">spike_in_amount</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;path/to/spike-in-rna.csv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">squeeze</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">dataset</span><span class="o">.</span><span class="n">add_spike_in</span><span class="p">(</span>
    <span class="n">spike_in_seq</span><span class="o">=</span><span class="s1">&#39;AAAAACAAAAACAAAAACAAA&#39;</span><span class="p">,</span>
    <span class="n">spike_in_amount</span><span class="o">=</span><span class="n">spike_in_amount</span><span class="p">,</span>
    <span class="n">base_table</span><span class="o">=</span><span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">original</span><span class="p">,</span>
    <span class="n">dist_type</span><span class="o">=</span><span class="s1">&#39;edit&#39;</span><span class="p">,</span>
    <span class="n">radius</span><span class="o">=</span><span class="mi">2</span>
<span class="p">)</span>

<span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered_ng_spikein</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">spike_in</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<p>The value in the table <code class="docutils literal notranslate"><span class="pre">filtered_ng</span></code> and <code class="docutils literal notranslate"><span class="pre">filtered_ng_spikin</span></code> are sequence mass with unit ‘ng’.</p>
</div>
<div class="section" id="calculate-the-dependent-variable">
<h3>Calculate the dependent variable<a class="headerlink" href="#calculate-the-dependent-variable" title="Permalink to this headline">¶</a></h3>
<p>The final step in the preprocessing is to calculate the dependent variable (y) in the kinetic model, in our case,
the fraction of sequences reacted. Given sequences have been quantify (as ng), the fraction reacted is the amount of
sequence in the reacted sample divided by the amount of sequence in the input sample. While this can be done by
regular <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> arithmetics, we also include
<code class="xref py py-class docutils literal notranslate"><span class="pre">ReactedFractionNormalizer</span></code> with additional functions (e.g.,
taking the median of multiple input replicates, see function documentation for details).</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">k_seq.data.transform</span> <span class="kn">import</span> <span class="n">ReactedFractionNormalizer</span>

<span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">reacted_fraction</span> <span class="o">=</span> <span class="n">ReactedFractionNormalizer</span><span class="p">(</span><span class="n">input_samples</span><span class="o">=</span><span class="s1">&#39;R0&#39;</span><span class="p">)(</span><span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered_ng</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<p>The table <code class="docutils literal notranslate"><span class="pre">reacted_fraction</span></code> contains the fractions reacted in samples for each sequences to fit
kinetic models.</p>
</div>
</div>
<div class="section" id="kinetic-model-fitting">
<h2>Kinetic model fitting<a class="headerlink" href="#kinetic-model-fitting" title="Permalink to this headline">¶</a></h2>
<div class="section" id="least-squares-fitting-with-bootstrapping">
<h3>Least-squares fitting with bootstrapping<a class="headerlink" href="#least-squares-fitting-with-bootstrapping" title="Permalink to this headline">¶</a></h3>
<p>For each sequence, we perform least-squares fitting to the kinetic model using
<a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.curve_fit.html">curve_fit</a>
function from <code class="docutils literal notranslate"><span class="pre">scipy</span></code> package.
Several bootstrapping schemas are implemented to robustly quantify the estimation uncertainty. All these functions
are enclosed in the class <a class="reference internal" href="k_seq.estimate.html#id58" title="k_seq.estimate.least_squares_batch.BatchFitter"><code class="xref py py-class docutils literal notranslate"><span class="pre">BatchFitter</span></code></a>.</p>
<p>Fitting thousands to hundreds of thousands sequence from <em>k</em>-Seq experimental could be computationally heavy, especially
with bootstrapping uncertainty quantify. Several optimizations are included in <code class="docutils literal notranslate"><span class="pre">BatchFitter</span></code> for large datasets:</p>
<ol class="arabic simple">
<li><p>Count reads are discrete numbers and lower abundant sequences are likely to contain same number of reads in all
samples. Such sequences containing same data from <em>k</em>-Seq are deduplicated through hashing and only one fitting
are performed for them.</p></li>
<li><p>During the fitting of a large dataset, results are streamed to disk for each sequence. If the fitting were interrupted,
it can restart by scanning the output folder and continue fitting only unfitted sequences.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">BatchFitter</span></code> perform fittings in parallel through multi-threading. Large jobs can be deployed to a High Performance
Computing (HPC) node with multi-cores for faster fitting.</p></li>
</ol>
<p>In this example, we fit sequences in the table <code class="docutils literal notranslate"><span class="pre">reacted_fraction</span></code> to the pseudo-first-order kinetics. The kinetic model
can be defined using python function following the format of <code class="docutils literal notranslate"><span class="pre">curve_fit</span></code>. The function has the independent variable
as the first argument and model parameters as the remaining arguments. Any model can be represented in such format can
be passed to <code class="docutils literal notranslate"><span class="pre">BatchFitter</span></code>. The fitter can perform single point estimation using all available data points and/or fit
bootstrapped data points for multiple times to estimate uncertainty.
In addition to model parameters <em>k</em> and <em>A</em>, we also intended to calculate the
product <em>kA</em> from each fitting results in point estimation and bootstrapping. A dictionary of functions (here <code class="docutils literal notranslate"><span class="pre">kA</span></code>)
can be passed to argument <code class="docutils literal notranslate"><span class="pre">metric</span></code> to calculate and save additional metrics after each fitting.
These functions should take a single argument containing a list of fitted model parameters (ordered according to
model function arguments). Please see below example and
<a class="reference internal" href="k_seq.estimate.html#id58" title="k_seq.estimate.least_squares_batch.BatchFitter"><code class="xref py py-class docutils literal notranslate"><span class="pre">BatchFitter</span></code></a>
for more details on fitting configuration.</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">k_seq.estimate.least_squares_batch</span> <span class="kn">import</span> <span class="n">BatchFitter</span>

<span class="k">def</span> <span class="nf">pseudo_first_order</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">A</span><span class="p">):</span>
    <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.479</span>
    <span class="n">t</span> <span class="o">=</span> <span class="mi">90</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">A</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span> <span class="n">alpha</span> <span class="o">*</span> <span class="n">t</span> <span class="o">*</span> <span class="n">k</span> <span class="o">*</span> <span class="n">c</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">y</span>

<span class="k">def</span> <span class="nf">kA</span><span class="p">(</span><span class="n">param</span><span class="p">):</span>
    <span class="n">k</span><span class="p">,</span> <span class="n">A</span> <span class="o">=</span> <span class="n">param</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">param</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">k</span> <span class="o">*</span> <span class="n">A</span>

<span class="n">fitter</span> <span class="o">=</span> <span class="n">BatchFitter</span><span class="p">(</span>
    <span class="n">y_dataframe</span><span class="o">=</span><span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">reacted_fraction</span><span class="p">,</span>
    <span class="n">x_data</span><span class="o">=</span><span class="n">dataset</span><span class="o">.</span><span class="n">x_values</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">pseudo_first_order</span><span class="p">,</span>
    <span class="n">x_label</span><span class="o">=</span><span class="s1">&#39;Substrate (M)&#39;</span><span class="p">,</span>         <span class="c1"># name for x values</span>
    <span class="n">y_label</span><span class="o">=</span><span class="s1">&#39;Fraction reacted&#39;</span><span class="p">,</span>      <span class="c1"># name for y values</span>
    <span class="n">bounds</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="mi">1</span><span class="p">)],</span>    <span class="c1"># value bound for fitting, format: [(k_min, A_min), (k_max, A_max)]</span>
    <span class="n">metrics</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;kA&#39;</span><span class="p">:</span> <span class="n">kA</span><span class="p">},</span>              <span class="c1"># additional metrics to calculate after each fitting</span>
    <span class="n">bootstrap_num</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>                <span class="c1"># bootstrap for 10 times (10 is for test purpose, recommend &gt; 100)</span>
    <span class="n">bs_record_num</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>                 <span class="c1"># save only 5 bootstrap fitting results to save disk space</span>
    <span class="n">bs_method</span><span class="o">=</span><span class="s1">&#39;data&#39;</span><span class="p">,</span>                <span class="c1"># bootstrap by resampling (x, y) data points</span>
    <span class="n">large_dataset</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>              <span class="c1"># if True, stream results to save_to folder during fitting</span>
    <span class="n">save_to</span><span class="o">=</span><span class="n">FIG_OUTPUT_DIR</span><span class="o">/</span><span class="s1">&#39;fitting&#39;</span><span class="p">,</span>
    <span class="n">rnd_seed</span><span class="o">=</span><span class="mi">23</span>
<span class="p">)</span>

<span class="n">fitter</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span>
    <span class="n">parallel_cores</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>               <span class="c1"># run fitting in parallel</span>
    <span class="n">point_estimate</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>             <span class="c1"># run point estimation using all data points</span>
    <span class="n">bootstrap</span><span class="o">=</span><span class="kc">True</span>                   <span class="c1"># run bootstrapping with configuration specified above</span>
<span class="p">)</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="fitting-results">
<h3>Fitting results<a class="headerlink" href="#fitting-results" title="Permalink to this headline">¶</a></h3>
<p>All fitting results are saved to the directory specified by the <code class="docutils literal notranslate"><span class="pre">save_to</span></code> arguments, a typical output folder contains
following content:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">summary.csv</span></code>: a table contain summarized output from fitting. In out example, it includes columns of sequence,
results from point estimation (<em>k</em>, <em>A</em>, and <em>kA</em> value), results from bootstrapping (mean, 2.5%, 50% or median, 97.5% of
estimated <em>k</em>, <em>A</em>, and <em>kA</em> values, with prefix ‘bs_’ )</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">seqs/</span></code> folder contains fitting details:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">seq_to_hash.json</span></code>: JSON file contains sequence to hash mapping</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">[hash_value].json</span></code>: list of JSON files contain the fitting details for hashed sequences, including data,
fitting configuration, point estimation results, bootstrap (uncertainty) results with selected records from
bootstrapped samples.</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>Results are also accessible in Python: <code class="docutils literal notranslate"><span class="pre">fitter.summary()</span></code> returns the fitting summary and fitting details are available
under <code class="docutils literal notranslate"><span class="pre">fitter.results</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This tutorial is under working progress and is frequently updated</p>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="k-seq_package.html" class="btn btn-neutral float-right" title="k-seq package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2021, Yuning Shen

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>