

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Getting Started Tutorial &mdash; k-seq 0.4.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/my_theme.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="k-seq package" href="k-seq_package.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> k-seq
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Getting Started Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data-preprocessing">Data preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#join-paired-end-reads-using-easydiver">Join paired-end reads using <code class="docutils literal notranslate"><span class="pre">EasyDIVER</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#load-count-files">Load count files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filter-sequences-and-samples">Filter sequences and samples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sequence-quantification">Sequence quantification</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#quantify-the-sequence">Quantify the sequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="#calculate-the-dependent-variable">Calculate the dependent variable</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#model-fitting-with-bootstrapping">Model fitting with bootstrapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="#additional-analysis">Additional analysis</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="k-seq_package.html"><code class="docutils literal notranslate"><span class="pre">k-seq</span></code> package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">k-seq</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Getting Started Tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/getting_started.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="getting-started-tutorial">
<h1>Getting Started Tutorial<a class="headerlink" href="#getting-started-tutorial" title="Permalink to this headline">¶</a></h1>
<p>This is a quick tutorial on using <code class="docutils literal notranslate"><span class="pre">k-seq</span></code> package to analyze <em>k</em>-Seq experiments data. The standard pipeline includes</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="#data-preprocessing">Data preprocessing</a></p></li>
<li><p><a class="reference internal" href="#sequence-quantification">Sequence quantification</a></p></li>
<li><p><a class="reference internal" href="#model-fitting-with-bootstrapping">Model fitting with bootstrapping</a></p></li>
</ol>
<p>Some <a class="reference internal" href="#additional-analysis">additional analysis</a> tools are also introduced to help design your experiments and characterizing your data.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="installation.html"><span class="doc">Installation</span></a> for setting up <cite>k-seq</cite> environment and installing <cite>k-seq</cite> package.
<a class="reference internal" href="k-seq_package.html"><span class="doc">k-seq package</span></a> for the API documentation of the package.</p>
</div>
<div class="section" id="data-preprocessing">
<h2>Data preprocessing<a class="headerlink" href="#data-preprocessing" title="Permalink to this headline">¶</a></h2>
<div class="section" id="join-paired-end-reads-using-easydiver">
<h3>Join paired-end reads using <code class="docutils literal notranslate"><span class="pre">EasyDIVER</span></code><a class="headerlink" href="#join-paired-end-reads-using-easydiver" title="Permalink to this headline">¶</a></h3>
<p>We integrated the <a class="reference external" href="https://github.com/ichen-lab-ucsb/EasyDIVER">EasyDIVER</a> pipeline in the script <code class="docutils literal notranslate"><span class="pre">fastq-to-counts.sh</span></code>
to perform paired-end joining for FASTQ files (<code class="docutils literal notranslate"><span class="pre">.fastq.gz</span></code>) from Illumina sequencing and deduplicated joined reads into counts
(number of reads) for unique sequences.</p>
<p>All FASTQ files should be organized in a folder with the format of file name: <code class="docutils literal notranslate"><span class="pre">[sample-name]_L[lane</span> <span class="pre">number]_R[1/2]_001.fastq.gz</span></code>.
<code class="docutils literal notranslate"><span class="pre">sample-name</span></code> is the string identifier for a sample so that reads from different sequencing flow-cell lanes can be pooled
together and <code class="docutils literal notranslate"><span class="pre">lane</span> <span class="pre">number</span></code> indicate the lane of the flow-cell. <code class="docutils literal notranslate"><span class="pre">R[1/2]</span></code> indicate the direction of paired-end reads
(<code class="docutils literal notranslate"><span class="pre">R1</span></code> for forward reads and <code class="docutils literal notranslate"><span class="pre">R2</span></code> for reverse reads) and each sample-lane expects to have matched files for both
directions.</p>
<p>Here is an example of input file folder including the paired-end reads from two samples <code class="docutils literal notranslate"><span class="pre">R4A-0A_S7</span></code> and <code class="docutils literal notranslate"><span class="pre">R4B-0A_S21</span></code>
from 4 lanes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L001_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L001_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L001_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L001_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L002_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L002_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L002_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L002_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L003_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L003_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L003_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L003_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L004_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L004_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">R4A</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S7_L004_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">R4B</span><span class="o">-</span><span class="mi">0</span><span class="n">A_S21_L004_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>To run <code class="docutils literal notranslate"><span class="pre">EasyDIVER</span></code>:</p>
<div class="highlight-shell notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10</pre></div></td><td class="code"><div class="highlight"><pre><span></span>&gt; fastq-to-counts.sh <span class="se">\</span>
      -i /path/to/input/folder <span class="se">\</span>
      -o /path/to/output/folder <span class="se">\</span>
      -p CTACGAATTC <span class="se">\ </span>             <span class="c1"># forward adapter sequence for trimming</span>
      -q CTGCAGTGAA <span class="se">\ </span>             <span class="c1"># reverse adapter sequence for trimming</span>
      -c <span class="se">\ </span>                        <span class="c1"># use completely matching in joining and discard reads with error</span>
      -a <span class="se">\ </span>                        <span class="c1"># join the paired-end reads first before trimming the adapters for reads with heavily overlapped regions</span>
      -T <span class="m">12</span>                        <span class="c1"># number of threads to run in parallel for pandaSeq</span>
&gt; ls /path/to/output/folder
counts  fastas  fastqs  histos  log.txt
</pre></div>
</td></tr></table></div>
<p>The output folder contains following files:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">counts</span></code>: folder contain deduplicated count files named as <code class="docutils literal notranslate"><span class="pre">[sample-name]_counts.txt</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fastqs</span></code>: joined FASTQ files named as <code class="docutils literal notranslate"><span class="pre">[sample-name].joined.fastq</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fastas</span></code>: joined FASTA files named as <code class="docutils literal notranslate"><span class="pre">[sample-name].joined.fasta</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">histos</span></code>: files contain the joined reads length histogram statistics, named as <code class="docutils literal notranslate"><span class="pre">[sample-files]_counts_histo.txt</span></code></p></li>
</ul>
<p>The output are the deduplicated count files with format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">number</span> <span class="n">of</span> <span class="n">unique</span> <span class="n">sequences</span> <span class="o">=</span>     <span class="mi">981824</span>
<span class="n">total</span> <span class="n">number</span> <span class="n">of</span> <span class="n">molecules</span> <span class="o">=</span>    <span class="mi">10096876</span>

<span class="n">GGGGGGGGATTCATGACTATT</span>                                                                 <span class="mi">1</span>
<span class="n">GGGGGGGAGTAGGACTGCAAA</span>                                                                 <span class="mi">1</span>
<span class="n">GGGGGGGAAGACTCCGGAACG</span>                                                                 <span class="mi">1</span>
<span class="n">GGGGGGGAACGCATTTCACGG</span>                                                                 <span class="mi">1</span>
<span class="n">GGGGGGGACGTTCACCGGCAA</span>                                                                 <span class="mi">1</span>
<span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="load-count-files">
<h3>Load count files<a class="headerlink" href="#load-count-files" title="Permalink to this headline">¶</a></h3>
<p>We next load and parse count files to Python using <a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData" title="k_seq.data.seq_data.SeqData"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.SeqData</span></code></a>.
<a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData" title="k_seq.data.seq_data.SeqData"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqData</span></code></a> is the main class to store and manipulate <em>k-Seq</em> data.
<a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData.from_count_files" title="k_seq.data.seq_data.SeqData.from_count_files"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SeqData.from_count_files</span></code></a> can load preprocessed count files
in a count file folder and extract metadata from the file name.</p>
<p>Here is a short example to load all count files under the folder <code class="docutils literal notranslate"><span class="pre">count_files</span></code> where these count files are named with
a pattern <code class="docutils literal notranslate"><span class="pre">s-{exp_cond:</span> <span class="pre">A,</span> <span class="pre">B,</span> <span class="pre">C,</span> <span class="pre">...}{rep:</span> <span class="pre">1,</span> <span class="pre">2,</span> <span class="pre">3,</span> <span class="pre">...}-_S{sample_id}_counts.txt</span></code>. To extract such metadata
during the loading and name each sample as “exp_cond + exp”, we can use “{}” to label the position of metadata with name
(and type <code class="docutils literal notranslate"><span class="pre">int</span></code> or <code class="docutils literal notranslate"><span class="pre">float</span></code>) in the braces and use “[]” to define the substring as the sample name:</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">k_seq.data</span> <span class="kn">import</span> <span class="n">SeqData</span>

<span class="n">dataset</span> <span class="o">=</span> <span class="n">SeqData</span><span class="o">.</span><span class="n">from_count_files</span><span class="p">(</span>
    <span class="n">count_files</span><span class="o">=</span><span class="s1">&#39;path/to/count_files&#39;</span><span class="p">,</span>
    <span class="n">pattern_filter</span><span class="o">=</span><span class="s1">&#39;_counts.&#39;</span><span class="p">,</span>
    <span class="n">name_template</span><span class="o">=</span><span class="s1">&#39;[</span><span class="si">{cond}</span><span class="s1">{exp_rep, int}]_S{smpl, int}_counts.txt&#39;</span><span class="p">,</span>
    <span class="n">sort_by</span><span class="o">=</span><span class="s1">&#39;smpl&#39;</span><span class="p">,</span>
    <span class="n">x_values</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
    <span class="n">x_unit</span><span class="o">=</span><span class="s1">&#39;M&#39;</span><span class="p">,</span>
    <span class="n">input_sample_name</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;R0&#39;</span><span class="p">],</span>
    <span class="n">note</span><span class="o">=</span><span class="s1">&#39;Example data&#39;</span>
<span class="p">)</span>
</pre></div>
</td></tr></table></div>
<p>For the complete usage, please refer to the function documentation
<a class="reference internal" href="k_seq.data.html#k_seq.data.seq_data.SeqData.from_count_files" title="k_seq.data.seq_data.SeqData.from_count_files"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SeqData.from_count_files</span></code></a></p>
<p>The counts in all samples are pooled in to a count table accessible under <code class="code docutils literal notranslate"><span class="pre">dataset.table.original</span></code>.</p>
</div>
<div class="section" id="filter-sequences-and-samples">
<h3>Filter sequences and samples<a class="headerlink" href="#filter-sequences-and-samples" title="Permalink to this headline">¶</a></h3>
<p>The loaded dataset with count table in <code class="xref py py-class docutils literal notranslate"><span class="pre">SeqData</span></code> can be further filtered according to selected quality control
criteria. For example, sequences contain ambiguous nucleotides (‘N’) and with undesired length could be removed using
<code class="xref py py-class docutils literal notranslate"><span class="pre">NoAmbiguityFilter</span></code> and
<code class="xref py py-class docutils literal notranslate"><span class="pre">SeqLengthFilter</span></code>. Some sample failed in the quality control for the experiment
can also be removed by <code class="xref py py-class docutils literal notranslate"><span class="pre">SampleFilter</span></code>.</p>
<p>These tables can be save as processed tables under <code class="xref py py-class docutils literal notranslate"><span class="pre">SeqData.tables</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">k_seq.data.filters</span> <span class="kn">import</span> <span class="n">NoAmbiguityFilter</span><span class="p">,</span> <span class="n">SeqLengthFilter</span>

<span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">filtered</span> <span class="o">=</span> <span class="n">SeqLengthFilter</span><span class="p">(</span><span class="n">min_len</span><span class="o">=</span><span class="mi">19</span><span class="p">,</span> <span class="n">max_len</span><span class="o">=</span><span class="mi">23</span><span class="p">)(</span>
    <span class="n">NoAmbiguityFilter</span><span class="p">()(</span>
        <span class="n">dataset</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">original</span>
    <span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="sequence-quantification">
<h2>Sequence quantification<a class="headerlink" href="#sequence-quantification" title="Permalink to this headline">¶</a></h2>
<div class="section" id="quantify-the-sequence">
<h3>Quantify the sequence<a class="headerlink" href="#quantify-the-sequence" title="Permalink to this headline">¶</a></h3>
<p>Depending on the design of the <em>k</em>-Seq experiment design, different quantification method could be used. The <code class="docutils literal notranslate"><span class="pre">k-seq</span></code>
package implemented two quantification functions: 1) by total sequence (DNA/RNA/protein) amount and 2) by amount of
external standards (spike-in sequence).</p>
<p>Specifically,</p>
</div>
<div class="section" id="calculate-the-dependent-variable">
<h3>Calculate the dependent variable<a class="headerlink" href="#calculate-the-dependent-variable" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="model-fitting-with-bootstrapping">
<h2>Model fitting with bootstrapping<a class="headerlink" href="#model-fitting-with-bootstrapping" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="additional-analysis">
<h2>Additional analysis<a class="headerlink" href="#additional-analysis" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="k-seq_package.html" class="btn btn-neutral float-right" title="k-seq package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Yuning Shen

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>