

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>k_seq.data: data preprocessing &mdash; k-seq 0.4.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/my_theme.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="k_seq.model: kinetics and pool models" href="k_seq.model.html" />
    <link rel="prev" title="k-seq package content" href="k-seq_package.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> k-seq
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started Tutorial</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="k-seq_package.html"><code class="docutils literal notranslate"><span class="pre">k-seq</span></code> package content</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">k_seq.data: data preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#k-seq-data-count-data">k_seq.data.count_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#k-seq-data-count-file">k_seq.data.count_file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-k_seq.data.datasets">k_seq.data.datasets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-k_seq.data.filters">k_seq.data.filters module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-k_seq.data.grouper">k_seq.data.grouper module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#k-seq-data-io-module">k_seq.data.io module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-k_seq.data.landscape">k_seq.data.landscape module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-k_seq.data.seq_data">k_seq.data.seq_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-k_seq.data.simu">k_seq.data.simu module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-k_seq.data.transform">k_seq.data.transform module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#k-seq-data-visualizer-module">k_seq.data.visualizer module</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="k_seq.model.html">k_seq.model: kinetics and pool models</a></li>
<li class="toctree-l2"><a class="reference internal" href="k_seq.estimate.html">k_seq.estimator: least squares fitting with bootstrapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="k_seq.utility.html">k_seq.utility</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">k-seq</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="k-seq_package.html"><code class="docutils literal notranslate"><span class="pre">k-seq</span></code> package content</a> &raquo;</li>
        
      <li>k_seq.data: data preprocessing</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/k_seq.data.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="k-seq-data-data-preprocessing">
<h1>k_seq.data: data preprocessing<a class="headerlink" href="#k-seq-data-data-preprocessing" title="Permalink to this headline">¶</a></h1>
<span class="target" id="module-k_seq.data"></span><p>This sub-package includes the modules for data handling, including:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">pre_processing</span></code>: core module in data pre-processing from count file to <em>SequenceSeq</em> for estimator</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">io</span></code>: module contains utility function for read, write and convert different file formats</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">analysis</span></code>: module contains functions for extra analysis for sequencing samples or reads to sample investigation
and sample pipeline quality control</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">simu</span></code>: module contains codes to generate simulated data used in analysis in paper (TODO: add paper citation)</p></li>
</ul>
<div class="section" id="k-seq-data-count-data">
<h2>k_seq.data.count_data<a class="headerlink" href="#k-seq-data-count-data" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="k-seq-data-count-file">
<h2>k_seq.data.count_file<a class="headerlink" href="#k-seq-data-count-file" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-k_seq.data.datasets">
<span id="k-seq-data-datasets"></span><h2>k_seq.data.datasets<a class="headerlink" href="#module-k_seq.data.datasets" title="Permalink to this headline">¶</a></h2>
<p>Pipeline for generating datasets for k-seq project
Available datasets:</p>
<blockquote>
<div><ul class="simple">
<li><p>BYO doped pool: byo-doped</p></li>
<li><p>BYO selection pool: byo-selected</p></li>
</ul>
</div></blockquote>
<dl class="py function">
<dt id="k_seq.data.datasets.load_dataset">
<code class="sig-prename descclassname">k_seq.data.datasets.</code><code class="sig-name descname">load_dataset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dataset</span></em>, <em class="sig-param"><span class="n">from_count_file</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.datasets.load_dataset" title="Permalink to this definition">¶</a></dt>
<dd><p>Load default dataset
Available dataset:</p>
<blockquote>
<div><ul class="simple">
<li><p>BYO-doped: ‘byo-doped’</p></li>
<li><p>BYO-selected: ‘byo-selected’</p></li>
<li><p>BFO: not implemented</p></li>
</ul>
</div></blockquote>
</dd></dl>

</div>
<div class="section" id="module-k_seq.data.filters">
<span id="k-seq-data-filters-module"></span><h2>k_seq.data.filters module<a class="headerlink" href="#module-k_seq.data.filters" title="Permalink to this headline">¶</a></h2>
<p>This module contains filters to apply on SeqData
TODO: simplify code
TODO: write test cases for filters</p>
<dl class="py class">
<dt id="k_seq.data.filters.CustomizedFilter">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.filters.</code><code class="sig-name descname">CustomizedFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mask_func</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.CustomizedFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.filters.Filter" title="k_seq.data.filters.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.filters.Filter</span></code></a></p>
<dl class="py method">
<dt id="k_seq.data.filters.CustomizedFilter.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mask_func</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.CustomizedFilter.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a Filter object from a given filter function
filter function should take a pd.DataFrame and return a boolean mask (True for pass, False to non pass)</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.CustomizedFilter.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.CustomizedFilter.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Implementation required
Return the the boolean mask for given seq_table
Wrapper over _get_mask to for preprocessing arguments
This content is a minimal example</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.filters.DetectedTimesFilter">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.filters.</code><code class="sig-name descname">DetectedTimesFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_detected_times</span><span class="o">=</span><span class="default_value">6</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.DetectedTimesFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.filters.Filter" title="k_seq.data.filters.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.filters.Filter</span></code></a></p>
<p>Filter sequences by the minimal times being detected (count &gt; 0)</p>
<dl class="py method">
<dt id="k_seq.data.filters.DetectedTimesFilter.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_detected_times</span><span class="o">=</span><span class="default_value">6</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.DetectedTimesFilter.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.DetectedTimesFilter.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_detected_times</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.DetectedTimesFilter.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Implementation required
Return the the boolean mask for given seq_table
Wrapper over _get_mask to for preprocessing arguments
This content is a minimal example</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.filters.Filter">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.filters.</code><code class="sig-name descname">Filter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.Filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">abc.ABC</span></code></p>
<p>Template class for filters</p>
<dl>
<dt><code class="docutils literal notranslate"><span class="pre">Filters</span></code> filter a <code class="docutils literal notranslate"><span class="pre">target</span></code> seq_table by index or column to another seq_table, similar to <code class="docutils literal notranslate"><span class="pre">Transformer</span></code>, but by</dt><dd><p>design do not change the content in the seq_table. To write a filter, one should:</p>
<ul class="simple">
<li><p>indicate default <code class="docutils literal notranslate"><span class="pre">axis</span></code> in <code class="docutils literal notranslate"><span class="pre">__init__</span></code> and store as an attribute, indicating which axis you are filtering,</p></li>
</ul>
<p>0 for index (row), 1 for columns</p>
<ul class="simple">
<li><p>implement staticmethod <code class="docutils literal notranslate"><span class="pre">_get_mask</span></code> with input of target seq_table and return a bool mask along the axis it</p></li>
</ul>
<p>filters. True for pass, False for not pass.
classmethod <code class="docutils literal notranslate"><span class="pre">filter</span></code> should then apply to a target pd.DataFrame seq_table and return a filtered seq_table</p>
<ul class="simple">
<li><p>implement method <code class="docutils literal notranslate"><span class="pre">get_mask</span></code> wrapper over <code class="docutils literal notranslate"><span class="pre">_get_mask</span></code> to assign stored attributes as argument. By design,</p></li>
</ul>
<p>it should only need to take a <code class="docutils literal notranslate"><span class="pre">target</span></code> argument, or without argument, returns <code class="docutils literal notranslate"><span class="pre">seq_data.mask</span></code> corresponding to
<code class="docutils literal notranslate"><span class="pre">seq_data.target</span></code></p>
</dd>
<dt>Note: please assign <code class="docutils literal notranslate"><span class="pre">target</span></code> at the end of __init__ as assigning <code class="docutils literal notranslate"><span class="pre">target</span></code> will do filtering and update</dt><dd><p><code class="docutils literal notranslate"><span class="pre">self.mask</span></code></p>
</dd>
</dl>
<dl class="py method">
<dt id="k_seq.data.filters.Filter.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.Filter.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>A filter should contains at least axis info for filtering</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.Filter.filter">
<em class="property">classmethod </em><code class="sig-name descname">filter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">remove_empty</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">reverse</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.Filter.filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Classmethod to directly apply filters</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.Filter.get_filtered_table">
<code class="sig-name descname">get_filtered_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">remove_empty</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">reverse</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.Filter.get_filtered_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a filtered seq_table</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target</strong> (<em>pd.DataFrame</em>) – seq_table to filter</p></li>
<li><p><strong>remove_empty</strong> (<em>bool</em>) – if remove all-zero items from another axis after filtering. Default False.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – if return filtered items instead of items passed the filter. Default False.</p></li>
<li><p><strong>axis</strong> (<em>0</em><em> or </em><em>1</em>) – if apply filter on index/row (0) or columns (1)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.Filter.get_mask">
<em class="property">abstract </em><code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.Filter.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Implementation required
Return the the boolean mask for given seq_table
Wrapper over _get_mask to for preprocessing arguments
This content is a minimal example</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.Filter.get_passed_item">
<code class="sig-name descname">get_passed_item</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">reverse</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.Filter.get_passed_item" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the items that pass the filter, pre-saved mask could be provided
if <code class="docutils literal notranslate"><span class="pre">reverse</span></code> is false, get filtered items</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.Filter.summary">
<code class="sig-name descname">summary</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.Filter.summary" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a pd.DataFrame as the summary</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.Filter.target">
<em class="property">property </em><code class="sig-name descname">target</code><a class="headerlink" href="#k_seq.data.filters.Filter.target" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.filters.FilterPipe">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.filters.</code><code class="sig-name descname">FilterPipe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filters</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.FilterPipe" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.filters.Filter" title="k_seq.data.filters.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.filters.Filter</span></code></a></p>
<p>Applies a collection of filters to the object in sequence, Not implemented yet</p>
<dl class="py method">
<dt id="k_seq.data.filters.FilterPipe.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.FilterPipe.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Implementation required
Return the the boolean mask for given seq_table
Wrapper over _get_mask to for preprocessing arguments
This content is a minimal example</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.filters.NoAmbiguityFilter">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.filters.</code><code class="sig-name descname">NoAmbiguityFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.NoAmbiguityFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.filters.Filter" title="k_seq.data.filters.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.filters.Filter</span></code></a></p>
<p>Filter out sequence with ambiguity nucleotides: N</p>
<dl class="py method">
<dt id="k_seq.data.filters.NoAmbiguityFilter.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.NoAmbiguityFilter.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Implementation required
Return the the boolean mask for given seq_table
Wrapper over _get_mask to for preprocessing arguments
This content is a minimal example</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.filters.PeakFilter">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.filters.</code><code class="sig-name descname">PeakFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">max_dist</span></em>, <em class="sig-param"><span class="n">dist_to_center</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">peak</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.PeakFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.filters.Filter" title="k_seq.data.filters.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.filters.Filter</span></code></a></p>
<p>Filter sequences by the provided distance to the center</p>
<dl class="py method">
<dt id="k_seq.data.filters.PeakFilter.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">max_dist</span></em>, <em class="sig-param"><span class="n">dist_to_center</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">peak</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.PeakFilter.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>dist_to_center or peak needed</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.PeakFilter.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_dist</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.PeakFilter.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Implementation required
Return the the boolean mask for given seq_table
Wrapper over _get_mask to for preprocessing arguments
This content is a minimal example</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.filters.SampleFilter">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.filters.</code><code class="sig-name descname">SampleFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">samples_to_keep</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">samples_to_remove</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.SampleFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.filters.Filter" title="k_seq.data.filters.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.filters.Filter</span></code></a></p>
<p>Filter samples based on index name</p>
<dl class="py method">
<dt id="k_seq.data.filters.SampleFilter.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">samples_to_keep</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">samples_to_remove</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.SampleFilter.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Implementation required
Return the the boolean mask for given seq_table
Wrapper over _get_mask to for preprocessing arguments
This content is a minimal example</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.filters.SeqLengthFilter">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.filters.</code><code class="sig-name descname">SeqLengthFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_len</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_len</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.SeqLengthFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.filters.Filter" title="k_seq.data.filters.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.filters.Filter</span></code></a></p>
<p>Filter out sequences with length &lt; min_len or &gt; max_len</p>
<dl class="py method">
<dt id="k_seq.data.filters.SeqLengthFilter.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_len</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_len</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.SeqLengthFilter.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Implementation required
Return the the boolean mask for given seq_table
Wrapper over _get_mask to for preprocessing arguments
This content is a minimal example</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.filters.SingletonFilter">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.filters.</code><code class="sig-name descname">SingletonFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.SingletonFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.filters.Filter" title="k_seq.data.filters.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.filters.Filter</span></code></a></p>
<p>Filter out sequence has only been detected 1 count across all samples</p>
<dl class="py method">
<dt id="k_seq.data.filters.SingletonFilter.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.SingletonFilter.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Implementation required
Return the the boolean mask for given seq_table
Wrapper over _get_mask to for preprocessing arguments
This content is a minimal example</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.filters.SpikeInFilter">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.filters.</code><code class="sig-name descname">SpikeInFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em>, <em class="sig-param"><span class="n">center_seq</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">radius</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dist_type</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">reverse</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.SpikeInFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.filters.Filter" title="k_seq.data.filters.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.filters.Filter</span></code></a></p>
<p>Filter out seqs that are spike-in</p>
<dl class="py attribute">
<dt id="k_seq.data.filters.SpikeInFilter.peak">
<code class="sig-name descname">peak</code><a class="headerlink" href="#k_seq.data.filters.SpikeInFilter.peak" title="Permalink to this definition">¶</a></dt>
<dd><p>the spike-in peak instance</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>landscape.Peak</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.SpikeInFilter.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em>, <em class="sig-param"><span class="n">center_seq</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">radius</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dist_type</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">reverse</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.SpikeInFilter.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target</strong> (<em>pd.DataFrame</em><em> or </em><em>seq_table.SeqData</em>) – needed to calculate distance of sequences to center seq
If target is pd.DataFrame, center, radius, and dist_type must
provide. If target is SeqData, it must infer from <code class="docutils literal notranslate"><span class="pre">SeqData.spike_in</span></code> accessor if applicable for
consistency</p></li>
<li><p><strong>center_seq</strong> (<em>str</em>) – center sequence of added spike-in sequence</p></li>
<li><p><strong>radius</strong> (<em>int</em>) – radius of spike-in peak. Seqs with distance ≤ radius will be filtered</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.filters.SpikeInFilter.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.filters.SpikeInFilter.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply to a target seq_table</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-k_seq.data.grouper">
<span id="k-seq-data-grouper-module"></span><h2>k_seq.data.grouper module<a class="headerlink" href="#module-k_seq.data.grouper" title="Permalink to this headline">¶</a></h2>
<p>Groupers slice seq_table into groups. E.g. input samples, reacted samples, different concentrations
TODO: simplify code</p>
<dl class="py class">
<dt id="k_seq.data.grouper.Grouper">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.grouper.</code><code class="sig-name descname">Grouper</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">group</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.grouper.Grouper" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Grouper of samples/sequences</p>
<dl class="simple">
<dt>Two types of grouper accepted:</dt><dd><p>Type 0 (list): initialize with group as list-like. This defines a single set of samples/sequences
Type 1 (dict): initialize with group as dict. This defines a collection of groups of samples/sequences</p>
</dd>
</dl>
<dl class="py attribute">
<dt id="k_seq.data.grouper.Grouper.target">
<code class="sig-name descname">target</code><a class="headerlink" href="#k_seq.data.grouper.Grouper.target" title="Permalink to this definition">¶</a></dt>
<dd><p>accessor for seq_table to group</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="k_seq.data.grouper.Grouper.axis">
<code class="sig-name descname">axis</code><a class="headerlink" href="#k_seq.data.grouper.Grouper.axis" title="Permalink to this definition">¶</a></dt>
<dd><p>axis to apply grouping (0 for index, 1 for columns)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>0 or 1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="k_seq.data.grouper.Grouper.group">
<code class="sig-name descname">group</code><a class="headerlink" href="#k_seq.data.grouper.Grouper.group" title="Permalink to this definition">¶</a></dt>
<dd><p>dictionary with structure {group_name: group_members}</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list or dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="k_seq.data.grouper.Grouper.type">
<code class="sig-name descname">type</code><a class="headerlink" href="#k_seq.data.grouper.Grouper.type" title="Permalink to this definition">¶</a></dt>
<dd><p>type of the grouper</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>0 or 1</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.grouper.Grouper.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">group</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.grouper.Grouper.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize a Grouper instance
:param group: list creates a Type 0 Grouper (single group) and dict creates a Type 1 Grouper</p>
<blockquote>
<div><p>(multiple groups)</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target</strong> (<em>pd.DataFrame</em>) – optional, target seq_table</p></li>
<li><p><strong>axis</strong> (<em>0</em><em> or </em><em>1</em>) – axis to apply the grouper</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.grouper.Grouper.get_table">
<code class="sig-name descname">get_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">group</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">remove_zero</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.grouper.Grouper.get_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a sub-seq_table from target given group</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.grouper.Grouper.split">
<code class="sig-name descname">split</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">remove_zero</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.grouper.Grouper.split" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.grouper.GrouperCollection">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.grouper.</code><code class="sig-name descname">GrouperCollection</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.grouper.GrouperCollection" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="k_seq.utility.html#k_seq.utility.func_tools.AttrScope" title="k_seq.utility.func_tools.AttrScope"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.utility.func_tools.AttrScope</span></code></a></p>
<p>A collection of groupers</p>
<dl class="py method">
<dt id="k_seq.data.grouper.GrouperCollection.add">
<code class="sig-name descname">add</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.grouper.GrouperCollection.add" title="Permalink to this definition">¶</a></dt>
<dd><p>add a grouper</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="k_seq.data.grouper.get_group">
<code class="sig-prename descclassname">k_seq.data.grouper.</code><code class="sig-name descname">get_group</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">table</span></em>, <em class="sig-param"><span class="n">group</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">remove_empty</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.grouper.get_group" title="Permalink to this definition">¶</a></dt>
<dd><p>Slice or split the table based on the group
:param table: target table to group
:type table: pd.DataFrame or SeqTable
:param group: single group or a set of groups
:type group: list-like or dict of list</p>
</dd></dl>

</div>
<div class="section" id="k-seq-data-io-module">
<h2>k_seq.data.io module<a class="headerlink" href="#k-seq-data-io-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-k_seq.data.landscape">
<span id="k-seq-data-landscape-module"></span><h2>k_seq.data.landscape module<a class="headerlink" href="#module-k_seq.data.landscape" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-k_seq.data.seq_data">
<span id="k-seq-data-seq-data"></span><h2>k_seq.data.seq_data<a class="headerlink" href="#module-k_seq.data.seq_data" title="Permalink to this headline">¶</a></h2>
<p>Submodule of <cite>SeqData</cite>, a rich functions class of seq_table for sequencing manipulation This module contains methods
for data pre-processing from count files to <code class="docutils literal notranslate"><span class="pre">CountFile</span></code> for estimator
For absolute quantification, it accepts absolute amount (e.g. measured by qPCR) or reacted fraction
.. todo:: - write output function for each class as JSON file and readable foler</p>
<dl class="py class">
<dt id="k_seq.data.seq_data.SeqData">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.seq_data.</code><code class="sig-name descname">SeqData</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">data_unit</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sample_list</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seq_list</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_note</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_sparse</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">seq_metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sample_metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">grouper</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">x_values</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">x_unit</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">note</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dataset_metadata</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Data instance to store k-seq result</p>
<dl class="py attribute">
<dt id="k_seq.data.seq_data.SeqData.seq_table">
<code class="sig-name descname">seq_table</code><a class="headerlink" href="#k_seq.data.seq_data.SeqData.seq_table" title="Permalink to this definition">¶</a></dt>
<dd><p>accessor for tables stored. Including <code class="docutils literal notranslate"><span class="pre">original</span></code> created during initialization.
tables stored should be <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> or <code class="docutils literal notranslate"><span class="pre">SeqTable</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="k_seq.utility.html#k_seq.utility.func_tools.AttrScope" title="k_seq.utility.func_tools.AttrScope">AttrScope</a></p>
</dd>
</dl>
</dd></dl>

<dl>
<dt>x_values (list-like, or dict): optional. value for controlled variables. If list-like, should have same length and order as samples; if dict, should have sample names as key</dt><dd><p>x_unit (str): optional. Unit for controlled variable. e.g. ‘uM’
seqs
samples
metadata (AttrScope): accessor for metadata for the dataset, includes</p>
<blockquote>
<div><p>sample (AttrScope): collection of metadata for samples if applicable
seq (AttrScope): collection of metadata for seqs if applicable
created_time (timestamp): datetime of the instance is created
note (str): note for the dataset
other dataset metadata objects could be added</p>
</div></blockquote>
<p>analysis (AttrScope): accessor to some pre-built analyses</p>
</dd>
<dt>Plugins:</dt><dd><p>grouper (GrouperCollection): collection of <code class="docutils literal notranslate"><span class="pre">Grouper</span></code> to slice subtables
spike_in (SpikeInNormalizer): optional. Accessor to the normalizer using spike-in
sample_total (TotalAmountNormalizer): optional. Accessor to the normalizer using total sample amount of seqs
analysis (SeqDataAnalyzer): built-in analysis tools to analyze SeqData object</p>
</dd>
</dl>
<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">data_unit</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sample_list</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seq_list</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_note</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_sparse</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">seq_metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sample_metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">grouper</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">x_values</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">x_unit</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">note</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dataset_metadata</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize a <cite>SeqData</cite> object</p>
<p>Args:
data (pd.DataFrame or np.ndarray): 2-D data with indices as sequences and columns as samples. If data is pd.DataFrame, values in index and column will be used as sequences and samples; if data is a 2-D np.ndarray, <cite>sample_list</cite> and <cite>seq_to_fit</cite> are needed with same length and order as data</p>
<blockquote>
<div><p>data_unit (str): The unit of seq_table values, e.g. counts, ng, M. Default counts.
sample_list (list-like): list of samples in the sample, should match the columns in the seq_table data
seq_to_fit:
data_note (str): Note for data seq_table
use_sparse (bool): If store the seq_table value as sparse matrix
seq_metadata (dict of objects): optional. Extra seq metadata
grouper:
x_values (list-like, or dict): optional. value for controlled variables. If list-like, should have same length and order as samples; if dict, should have sample names as key
x_unit (str): optional. Unit for controlled variable. e.g. ‘uM’
note (str): Note for dataset/seq_table
dataset_metadata (dict of objects): optional. Extra dataset metadata</p>
</div></blockquote>
<dl class="simple">
<dt>grouper (dict of list, or dict of dict of list): optional. dict of list (Type 1) of dict of list (Type 2) to</dt><dd><p>create grouper plugin</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.add_grouper">
<code class="sig-name descname">add_grouper</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData.add_grouper" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an accessor of GrouperCollection of SeqData if not yet. Add Groupers to the accessor
Initialize a Grouper instance with keyword arguments with a dictionary of:</p>
<blockquote>
<div><dl class="simple">
<dt>group (list or dict): list creates a Type 0 Grouper (single group) and dict creates a Type 1 Grouper</dt><dd><p>(multiple groups)</p>
</dd>
</dl>
<p>target (pd.DataFrame): optional, target table
axis (0 or 1): axis to apply the grouper</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.add_sample_total">
<code class="sig-name descname">add_sample_total</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">total_amounts</span></em>, <em class="sig-param"><span class="n">full_table</span></em>, <em class="sig-param"><span class="n">unit</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData.add_sample_total" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Add TotalAmountNormalizer to quantify sequences with their total amount in each sample</dt><dd><p>as <cite>sample_total</cite></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>total_amounts</strong> (<em>dict</em><em> or </em><em>pd.Series</em>) – Total DNA amount for samples measured in experiment</p></li>
<li><p><strong>full_table</strong> (<em>pd.DataFrame</em>) – seq_table where the total amount were measured and normalize to</p></li>
<li><p><strong>unit</strong> (<em>str</em>) – Unit of amount measured</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.add_spike_in">
<code class="sig-name descname">add_spike_in</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">base_table</span></em>, <em class="sig-param"><span class="n">spike_in_seq</span></em>, <em class="sig-param"><span class="n">spike_in_amount</span></em>, <em class="sig-param"><span class="n">radius</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">unit</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dist_type</span><span class="o">=</span><span class="default_value">'edit'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData.add_spike_in" title="Permalink to this definition">¶</a></dt>
<dd><p>Add SpikeInNormalizer to quantify seq amount using spike-in sequence as accessor
<code class="docutils literal notranslate"><span class="pre">spike_in</span></code> to the instance</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>base_table</strong> (<em>pd.DataFrame</em>) – base_table includes spike-in sequences to calculate normalization factor</p></li>
<li><p><strong>spike_in_seq</strong> (<em>str</em>) – center sequence for spike-in</p></li>
<li><p><strong>spike_in_amount</strong> (<em>list-like</em><em>, </em><em>dict</em><em>, or </em><em>pd.Series</em>) – added spike_in amount, dict and pd.Series should have key of samples in base_table list-like should have same length as number of samples (cols) in base_table</p></li>
<li><p><strong>radius</strong> (<em>int</em>) – Radius of spike-in peak, seqs less or equal to radius away from center are spike-in seqs</p></li>
<li><p><strong>unit</strong> (<em>str</em>) – unit of spike-in amount</p></li>
<li><p><strong>dist_type</strong> (<em>str</em>) – ‘edit’ or ‘hamming’ as distance measure. Default ‘edit’ to include insertion / deletion</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.from_count_files">
<em class="property">static </em><code class="sig-name descname">from_count_files</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData.from_count_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a <code class="docutils literal notranslate"><span class="pre">SeqData</span></code> instance from a folder of count files</p>
<blockquote>
<div><dl>
<dt>Args:</dt><dd><p>count_files (str): root directory to search for count files
file_list (list of str): optional, only includes the count files with names in the file_list
pattern_filter (str): optional, filter file names based on this pattern, wildcards <code class="docutils literal notranslate"><span class="pre">*/?</span></code> are allowed
black_list (list of str): optional, file names included in black_list will be excluded
name_template (str): naming convention to extract metadata. Use <code class="docutils literal notranslate"><span class="pre">[...]</span></code> to include the region of sample_name,</p>
<blockquote>
<div><p>use <code class="docutils literal notranslate"><span class="pre">{domain_name[,</span> <span class="pre">int/float]}</span></code> to indicate region of domain to extract as metadata, including
<code class="docutils literal notranslate"><span class="pre">int</span></code> or <code class="docutils literal notranslate"><span class="pre">float</span></code> will convert the domain value to int/float in applicable, otherwise, string</p>
</div></blockquote>
<p>sort_by (str): sort the order of samples based on given domain
dry_run (bool): only return the parsed count file names and metadata without actual reading in data</p>
</dd>
</dl>
<p>x_values (list-like, or dict): optional. value for controlled variables. If list-like, should have same length and order as samples; if dict, should have sample names as key
x_unit (str): optional. Unit for controlled variable. e.g. ‘uM’
input_sample_name (list of str): optional. Indicate input samples (unreacted)
sample_metadata (dict of objects): optional. Extra sample metadata
note (str): Note for dataset/seq_table</p>
</div></blockquote>
<p class="rubric">Example</p>
<p>Example on metadata extraction from pattern:
&gt;&gt;&gt; metadata = extract_metadata(</p>
<blockquote>
<div><blockquote>
<div><p>sample_name = “R4B-1250A_S16_counts.txt”
template = “R4[{exp_rep}-{concentration, float}{seq_rep}_S{id, int}]_counts.txt”</p>
</div></blockquote>
<p>)</p>
</div></blockquote>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">metadata</span>
<span class="go">{</span>
<span class="go">    &#39;name&#39;: &#39;B-1250A_S16&#39;,</span>
<span class="go">    &#39;exp_rep&#39;: &#39;B&#39;,</span>
<span class="go">    &#39;concentration&#39;: 1250.0,</span>
<span class="go">    &#39;seq_rep&#39;: &#39;A&#39;,</span>
<span class="go">    &#39;id&#39;: 16</span>
<span class="go">}</span>
</pre></div>
</div>
<dl>
<dt>Notice: two back-to-back domain can only be parsed if one of them is numeric and one of them is alphabetic, and missing</dt><dd><p>value will raise error</p>
<p>Valid: matching ‘-A1-‘ to ‘-{{sample}}{{replicate, int}}-‘ gives {{ ‘sample’: ‘A’, ‘replicate’: 1}}
Not valid: matching ‘-A-‘ to ‘-{{sample}}{{replicate, int}}-‘ will cause error</p>
<blockquote>
<div><p>matching ‘-AA-‘ to ‘-{{sample}}{{replicate}}-‘ will cause error</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.from_json">
<code class="sig-name descname">from_json</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData.from_json" title="Permalink to this definition">¶</a></dt>
<dd><p>TODO: add json</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.from_pickle">
<em class="property">static </em><code class="sig-name descname">from_pickle</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData.from_pickle" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.samples">
<em class="property">property </em><code class="sig-name descname">samples</code><a class="headerlink" href="#k_seq.data.seq_data.SeqData.samples" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.seqs">
<em class="property">property </em><code class="sig-name descname">seqs</code><a class="headerlink" href="#k_seq.data.seq_data.SeqData.seqs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.to_json">
<code class="sig-name descname">to_json</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData.to_json" title="Permalink to this definition">¶</a></dt>
<dd><p>More generalized JSON file
TODO: add to_json and from_json</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.to_pickle">
<code class="sig-name descname">to_pickle</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData.to_pickle" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqData.update_analysis">
<code class="sig-name descname">update_analysis</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqData.update_analysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Update accessor to SeqDataAnalyzer</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.seq_data.SeqTable">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.seq_data.</code><code class="sig-name descname">SeqTable</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.core.frame.DataFrame</span></code></p>
<p>Extended <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> with added property and functions for SeqTable</p>
<dl class="simple">
<dt>Additional Attributes:</dt><dd><p>unit (str): unit of entries in this seq_table
note (str): note for this seq_table
samples (pd.Series): samples in the seq_table
seqs (pd.Series): sequences in the seq_table</p>
</dd>
<dt>Additional Methods:</dt><dd><p>about: print a summary of the seq_table
analysis: accessor to SeqTableAnalyzer</p>
</dd>
</dl>
<dl class="py method">
<dt id="k_seq.data.seq_data.SeqTable.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqTable.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize SeqTable instance
Additional kwargs:
unit:</p>
<blockquote>
<div><p>note (str): Note for dataset/seq_table
use_sparse (bool): If store the seq_table value as sparse matrix</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqTable.about">
<code class="sig-name descname">about</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqTable.about" title="Permalink to this definition">¶</a></dt>
<dd><p>Quick view of SeqTable</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqTable.density">
<em class="property">property </em><code class="sig-name descname">density</code><a class="headerlink" href="#k_seq.data.seq_data.SeqTable.density" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqTable.describe">
<code class="sig-name descname">describe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentiles</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">include</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">exclude</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqTable.describe" title="Permalink to this definition">¶</a></dt>
<dd><p>return major stable statistics</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqTable.filter_axis">
<code class="sig-name descname">filter_axis</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filter</span></em>, <em class="sig-param"><span class="n">axis</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">remove_empty</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqTable.filter_axis" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter seq_table along with one axis</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filter</strong> (<em>callable</em>) – a callable to apply on row/columns and returns a bool value</p></li>
<li><p><strong>axis</strong> (<em>0</em><em> or </em><em>1</em>) – the axis to filter. 0: row, seqs; 1: column, sample</p></li>
<li><p><strong>remove_empty</strong> (<em>bool</em>) – If remove the empty column/rows after filtering</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – if change the seq_table inplace. If False, return a new seq_table</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqTable.samples">
<em class="property">property </em><code class="sig-name descname">samples</code><a class="headerlink" href="#k_seq.data.seq_data.SeqTable.samples" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqTable.seqs">
<em class="property">property </em><code class="sig-name descname">seqs</code><a class="headerlink" href="#k_seq.data.seq_data.SeqTable.seqs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="k_seq.data.seq_data.SeqTable.update_analysis">
<code class="sig-name descname">update_analysis</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.SeqTable.update_analysis" title="Permalink to this definition">¶</a></dt>
<dd><p>update the accessor to SeqTableAnalyzer</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="k_seq.data.seq_data.slice_table">
<code class="sig-prename descclassname">k_seq.data.seq_data.</code><code class="sig-name descname">slice_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">table</span></em>, <em class="sig-param"><span class="n">axis</span></em>, <em class="sig-param"><span class="n">keys</span></em>, <em class="sig-param"><span class="n">remove_empty</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.seq_data.slice_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Slice pd.DataFrame seq_table with a list of key values or filter functions returning True/False along
given axis. Optional to remove all zero entries
:param seq_table: seq_table to slice
:type seq_table: pd.DataFrame
:param keys: list of keys to preserve. If is callable, apply to row/column of seq_table and returns</p>
<blockquote>
<div><p>bool of preserve (True) or discard (False)</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>0</em><em> or </em><em>1</em>) – which axis to filter</p></li>
<li><p><strong>remove_empty</strong> (<em>bool</em>) – If remove the empty column/rows after filtering</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-k_seq.data.simu">
<span id="k-seq-data-simu-module"></span><h2>k_seq.data.simu module<a class="headerlink" href="#module-k_seq.data.simu" title="Permalink to this headline">¶</a></h2>
<p>Module contains code to simulate data</p>
<dl class="py class">
<dt id="k_seq.data.simu.DistGenerators">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.simu.</code><code class="sig-name descname">DistGenerators</code><a class="headerlink" href="#k_seq.data.simu.DistGenerators" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A collection of random value generators from commonly used distributions.
<cite>uniq_seq_num</cite> number of independent draw of distribution are returned</p>
<dl class="simple">
<dt>Available distributions:</dt><dd><p>lognormal
uniform
compo_lognormal</p>
</dd>
</dl>
<dl class="py method">
<dt id="k_seq.data.simu.DistGenerators.compo_lognormal">
<em class="property">static </em><code class="sig-name descname">compo_lognormal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">size</span></em>, <em class="sig-param"><span class="n">loc</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">scale</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">c95</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.DistGenerators.compo_lognormal" title="Permalink to this definition">¶</a></dt>
<dd><p>Sample a pool composition from a log-normal distribution
indicate with <cite>loc</cite> and <cite>scale</cite>, or <cite>c95</cite></p>
<p class="rubric">Example</p>
<p>scale = 0 means an evenly distributed pool with all components have relative abundance 1/uniq_seq_num</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size</strong> (<cite>int</cite>) – uniq_seq_num of the pool</p></li>
<li><p><strong>loc</strong> (<cite>float</cite>) – center of log-normal distribution</p></li>
<li><p><strong>scale</strong> (<cite>float</cite>) – log variance of the distribution</p></li>
<li><p><strong>c95</strong> ([<cite>float</cite>, <cite>float</cite>]) – 95% percentile of log-normal distribution</p></li>
<li><p><strong>seed</strong> – random seed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.simu.DistGenerators.lognormal">
<em class="property">static </em><code class="sig-name descname">lognormal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">loc</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">scale</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">c95</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.DistGenerators.lognormal" title="Permalink to this definition">¶</a></dt>
<dd><p>Sample from a log-normal distribution
indicate with <cite>loc</cite> and <cite>scale</cite>, or <cite>c95</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size</strong> (<cite>int</cite>) – number of values to draw</p></li>
<li><p><strong>loc</strong> (<cite>float</cite>) – center of log-normal distribution, default 0</p></li>
<li><p><strong>scale</strong> (<cite>float</cite>) – log variance of the distribution, default 0</p></li>
<li><p><strong>c95</strong> ([<cite>float</cite>, <cite>float</cite>]) – 95% percentile of log-normal distribution</p></li>
<li><p><strong>seed</strong> – random seed</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a draw from distribution with given uniq_seq_num</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.simu.DistGenerators.uniform">
<em class="property">static </em><code class="sig-name descname">uniform</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">low</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">high</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.DistGenerators.uniform" title="Permalink to this definition">¶</a></dt>
<dd><p>Sample from a uniform distribution</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.simu.PoolParamGenerator">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.simu.</code><code class="sig-name descname">PoolParamGenerator</code><a class="headerlink" href="#k_seq.data.simu.PoolParamGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Functions to generate parameters for a set of sequence in a sequence pool</p>
<dl class="py method">
<dt>
<code class="sig-name descname">- sample_from_iid_dist</code></dt>
<dd></dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">- sample_from_dataframe</code></dt>
<dd></dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>pd.DataFrame with columns of parameters and rows of simulated parameter for each sequence</p>
</dd>
</dl>
<dl class="py method">
<dt id="k_seq.data.simu.PoolParamGenerator.sample_from_dataframe">
<em class="property">classmethod </em><code class="sig-name descname">sample_from_dataframe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">df</span></em>, <em class="sig-param"><span class="n">uniq_seq_num</span></em>, <em class="sig-param"><span class="n">replace</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">weights</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.PoolParamGenerator.sample_from_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate parameter by resampling rows of a given dataframe
:param df: dataframe contains parameters as columns to sample from,</p>
<blockquote>
<div><p>needs to have <cite>p0</cite> as one column for heterogenous pool</p>
</div></blockquote>
<dl class="field-list simple">
</dl>
<p>uniq_seq_num (int): Number of unique sequences from simulation</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.simu.PoolParamGenerator.sample_from_iid_dist">
<em class="property">static </em><code class="sig-name descname">sample_from_iid_dist</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">uniq_seq_num</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">param_generators</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.PoolParamGenerator.sample_from_iid_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate the seq parameters from individual draws of distributions
Parameter:</p>
<blockquote>
<div><p>p0: initial fraction for each sequence for uneven pool
depending on the model. e.g. first-order model needs to include k and A</p>
</div></blockquote>
<dl class="simple">
<dt>Accepted parameter input:</dt><dd><ul class="simple">
<li><p>list-like: if uniq_seq_num does not match as expected uniq_seq_num, resample with replacement</p></li>
<li><p>generator: given generator returns a random parameter</p></li>
<li><p>callable: if taking <cite>uniq_seq_num</cite> as an argument, needs to return a uniq_seq_num vector of sampled parameter or a
generator to generate a uniq_seq_num vector; if not taking <cite>uniq_seq_num</cite> as an argument, needs to return single
sample</p></li>
</ul>
</dd>
</dl>
<p>Args:
p0 (list-like, generator, or callable): reserved argument for initial pool composition (fraction)</p>
<blockquote>
<div><p>uniq_seq_num (int): Number of unique sequences from simulation
seed (int): random seed for repeatability
param_generators (kwargs of list-like, generator, or callable): parameter generator depending on the model</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a n_row = uniq_seq_num pd.DataFrame contains generated parameters</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.simu.SimulationResults">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.simu.</code><code class="sig-name descname">SimulationResults</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dataset_dir</span></em>, <em class="sig-param"><span class="n">result_dir</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.SimulationResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to load simulation result</p>
<dl class="py method">
<dt id="k_seq.data.simu.SimulationResults.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dataset_dir</span></em>, <em class="sig-param"><span class="n">result_dir</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.SimulationResults.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Survey estimation results
- load fitting results from <cite>result_dir/fit_summary.csv</cite>
- load truth and input count infor from <cite>dataset_dir/truth.csv</cite> and <cite>input_counts</cite></p>
<p>Optional to include:
- input_counts: counts of sequences in the input pool
- mean_counts: mean counts in all samples (input and reacted)</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>table of estimated k, A, kA
truth: table of true k, A, p0, ka, and input_counts
seq_list: list of indices of sequences that were able to estimate</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>results</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.simu.SimulationResults.get_est_results">
<code class="sig-name descname">get_est_results</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">param</span></em>, <em class="sig-param"><span class="n">pred_type</span><span class="o">=</span><span class="default_value">'point_est'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.SimulationResults.get_est_results" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the estimation (pred) and truth of given parameter</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.simu.SimulationResults.get_fold_range">
<code class="sig-name descname">get_fold_range</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">param</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.SimulationResults.get_fold_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the ratio of 97.5-percentile to 2.5-percentile</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.simu.SimulationResults.get_uncertainty_accuracy">
<code class="sig-name descname">get_uncertainty_accuracy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">param</span></em>, <em class="sig-param"><span class="n">pred_type</span><span class="o">=</span><span class="default_value">'bs_ci95'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.SimulationResults.get_uncertainty_accuracy" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the accuracy of uncertainty estimation if uncertainty range includes the truth</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="k_seq.data.simu.get_pct_gaussian_error">
<code class="sig-prename descclassname">k_seq.data.simu.</code><code class="sig-name descname">get_pct_gaussian_error</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rate</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.get_pct_gaussian_error" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a function to apply Gaussian error proportional to the value</p>
</dd></dl>

<dl class="py function">
<dt id="k_seq.data.simu.simulate_counts">
<code class="sig-prename descclassname">k_seq.data.simu.</code><code class="sig-name descname">simulate_counts</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">uniq_seq_num</span></em>, <em class="sig-param"><span class="n">x_values</span></em>, <em class="sig-param"><span class="n">total_reads</span></em>, <em class="sig-param"><span class="n">p0_generator</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">kinetic_model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">count_model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">total_amount_error</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">param_sample_from_df</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">weights</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">replace</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">reps</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">note</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">save_to</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">param_generators</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.simulate_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate sequencing count dataset given kinetic and count model</p>
<dl>
<dt>Procedure:</dt><dd><ol class="arabic simple">
<li><p>parameter for each unique sequences were sampled from param_sample_from_df and kwargs</p></li>
</ol>
<blockquote>
<div><p>(param_generators). It is an even pool if p0 is not provided. No repeated parameters.</p>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><p>simulate the reacted amount / fraction of sequences with each controlled variable in x_values</p></li>
<li><p>Simulated counts with given total total_reads were simulated for input pool and reacted pools.</p></li>
</ol>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>uniq_seq_num</strong> (<em>int</em>) – Number of unique sequences from simulation</p></li>
<li><p><strong>x_values</strong> (<em>list-like</em>) – list of controlled variables in each experiment setup,negative value means it is an initial pool</p></li>
<li><p><strong>total_reads</strong> – </p></li>
<li><p><strong>p0</strong> (<em>list-like</em><em>, </em><em>generator</em><em>, or </em><em>callable</em>) – reserved argument for initial pool composition (fraction)</p></li>
<li><p><strong>kinetic_model</strong> (<em>callable</em>) – model the amount of sequences in reaction given input variables. Default <cite>BYOModel.amount_first_order</cite></p></li>
<li><p><strong>count_model</strong> (<em>callable</em>) – model the sequencing counts w.r.t. total reads and pool composition.Default MultiNomial model.</p></li>
<li><p><strong>param_sample_from_df</strong> (<em>pd.DataFrame</em>) – optional to sample sequences from given table</p></li>
<li><p><strong>weights</strong> (<em>list</em><em> or </em><em>str</em>) – weights/col of weight for sampling from table</p></li>
<li><p><strong>total_amount_error</strong> (<em>float</em><em> or </em><em>callable</em>) – float as the standard deviation for a fixed Gaussian error, or any error function on the DNA amount. Use 0 for no introduced error.</p></li>
<li><p><strong>conv_reps</strong> – </p></li>
<li><p><strong>seed</strong> (<em>int</em>) – random seed for repeatability</p></li>
<li><p><strong>save_to</strong> (<em>str</em>) – optional, path to save the simulation results with x, Y, truth csv file and a pickled SeqData object</p></li>
<li><p><strong>param_generator</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>c, n value for samples
Y (pd.DataFrame): simulated sequence counts for given samples
param_table (pd.DataFrame): seq_table list the parameters for simulated sequences
seq_table (data.SeqData): a SeqData object to stores all the data</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>x (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="k_seq.data.simu.simulate_on_byo_doped_condition_from_exp_results">
<code class="sig-prename descclassname">k_seq.data.simu.</code><code class="sig-name descname">simulate_on_byo_doped_condition_from_exp_results</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dataset</span></em>, <em class="sig-param"><span class="n">fitting_res</span></em>, <em class="sig-param"><span class="n">uniq_seq_num</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">x_values</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">total_reads</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sequencing_depth</span><span class="o">=</span><span class="default_value">40</span></em>, <em class="sig-param"><span class="n">n_input</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">table_name</span><span class="o">=</span><span class="default_value">'original'</span></em>, <em class="sig-param"><span class="n">total_dna_error_rate</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">23</span></em>, <em class="sig-param"><span class="n">plot_dist</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">save_to</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.simulate_on_byo_doped_condition_from_exp_results" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Simulate k-seq count dataset based on the experimental condition of BYO-doped pool, that</dt><dd><p>t: reaction time (90 min)
alpha: degradation ratio of BYO (0.479)
x_values: controlled BYO concentration points: 1 input pool with triple sequencing depth,</p>
<blockquote>
<div><dl class="simple">
<dt>5 BYO concentration with triplicates:</dt><dd><dl class="simple">
<dt>[-1 (input pool),</dt><dd><p>2e-6, 2e-6, 2e-6,
10e-6, 10e-6, 10e-6,
50e-6, 50e-6, 50e-6,
250e-6, 250e-6, 250e-6,
1260e-6, 1260e-6, 1260e-6]</p>
</dd>
</dl>
</dd>
</dl>
</div></blockquote>
</dd>
<dt>Parameter for each sequences were sampled from previous point estimate results:</dt><dd><ul class="simple">
<li><p>point_est_csv: load point estimates results to extract estimated k and A</p></li>
<li><p>seqtable_path: path to load input sample SeqData object to get p0 information</p></li>
</ul>
</dd>
</dl>
<p>Returns:
x (pd.DataFrame): controlled variable (c, n) for samples</p>
<blockquote>
<div><p>Y (pd.DataFrame): simulated sequence counts for given samples
param_table (pd.DataFrame): seq_table list the parameters for simulated sequences, including p0, k, A, kA
truth (pd.DataFrame): true values of parameters (e.g. p0, k, A) for simulated sequences
seq_table (data.SeqData): a SeqData object to stores all the data</p>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt id="k_seq.data.simu.simulate_w_byo_doped_condition_from_param_dist">
<code class="sig-prename descclassname">k_seq.data.simu.</code><code class="sig-name descname">simulate_w_byo_doped_condition_from_param_dist</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">uniq_seq_num</span></em>, <em class="sig-param"><span class="n">depth</span></em>, <em class="sig-param"><span class="n">p0_loc</span></em>, <em class="sig-param"><span class="n">p0_scale</span></em>, <em class="sig-param"><span class="n">k_95</span></em>, <em class="sig-param"><span class="n">total_dna_error_rate</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">23</span></em>, <em class="sig-param"><span class="n">save_to</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">plot_dist</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.simu.simulate_w_byo_doped_condition_from_param_dist" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Deprecated. Simulate k-seq count dataset similar to the experimental condition of BYO-doped pool, that</dt><dd><p>t: reaction time (90 min)
alpha: degradation ratio of BYO (0.479)
x_values: controlled BYO concentration points: 1 input pool with triple sequencing depth,</p>
<blockquote>
<div><dl class="simple">
<dt>5 BYO concentration with triplicates:</dt><dd><dl class="simple">
<dt>[-1 (input pool),</dt><dd><p>2e-6, 2e-6, 2e-6,
10e-6, 10e-6, 10e-6,
50e-6, 50e-6, 50e-6,
250e-6, 250e-6, 250e-6,
1260e-6, 1260e-6, 1260e-6]</p>
</dd>
</dl>
</dd>
</dl>
</div></blockquote>
</dd>
<dt>Parameter for each sequences were sampled from given distribution defined from arguments</dt><dd><ul class="simple">
<li><p>p0: log normal from exp(N(p0_loc, p0_scale))</p></li>
<li><p>k: log normal from k_95 95-percentile for k</p></li>
<li><p>A: uniform from [0, 1]</p></li>
</ul>
</dd>
</dl>
<p>Other args:
uniq_seq_num (int): Number of unique sequences from simulation</p>
<blockquote>
<div><p>depth (int or float): sequence depth defined on mean reads per sequence
total_amount_error (float or callable): float as the standard deviation for a fixed Gaussian error, or any error function on the DNA amount. Use 0 for no introduced error.
save_to (str): optional, path to save the simulation results with x, Y, truth csv file and a pickled SeqData object
plot_dist (bool): if pairwise figures of distribution for simulated parameters (p0, A, k, kA)</p>
</div></blockquote>
<p>Returns:
x (pd.DataFrame): controlled variable (c, n) for samples</p>
<blockquote>
<div><p>Y (pd.DataFrame): simulated sequence counts for given samples
param_table (pd.DataFrame): seq_table list the parameters for simulated sequences, including p0, k, A, kA
truth (pd.DataFrame): true values of parameters (e.g. p0, k, A) for simulated sequences
seq_table (data.SeqData): a SeqData object to stores all the data</p>
</div></blockquote>
</dd></dl>

</div>
<div class="section" id="module-k_seq.data.transform">
<span id="k-seq-data-transform-module"></span><h2>k_seq.data.transform module<a class="headerlink" href="#module-k_seq.data.transform" title="Permalink to this headline">¶</a></h2>
<dl class="simple">
<dt>Module contains the classes to transform seq_table (<cite>pd.DataFrame</cite> or <cite>SeqData</cite>) instance as well as related calculation</dt><dd><p>and visualizations</p>
</dd>
<dt>Current available transformers:</dt><dd><ul class="simple">
<li><p>SpikeInNormalizer: normalize by spike-in</p></li>
<li><p>DnaAmountNormalizer: normalize by total dna amount</p></li>
<li><p>ReactedFractionNormalizer: transform to relative abundance</p></li>
<li><p>BYOSelectedPoolNormalizerByAbe: curated quantification factor used by Abe</p></li>
</ul>
</dd>
</dl>
<dl class="py class">
<dt id="k_seq.data.transform.ReactedFractionNormalizer">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.transform.</code><code class="sig-name descname">ReactedFractionNormalizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">input_samples</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">reduce_method</span><span class="o">=</span><span class="default_value">'median'</span></em>, <em class="sig-param"><span class="n">remove_empty</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.ReactedFractionNormalizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.transform.Transformer" title="k_seq.data.transform.Transformer"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.transform.Transformer</span></code></a></p>
<p>Get reacted fraction of each sequence from an absolute amount seq_table</p>
<dl class="py method">
<dt id="k_seq.data.transform.ReactedFractionNormalizer.apply">
<code class="sig-name descname">apply</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">input_samples</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">reduce_method</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">remove_empty</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.ReactedFractionNormalizer.apply" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert absolute amount to reacted fraction
:param target: the seq_table with absolute amount to normalize on inputs, including input pools
:type target: pd.DataFrame
:param input_samples: list of indices of input pools
:type input_samples: list of str
:param reduce_method: ‘mean’ or ‘median’ or a callable apply on a pd.DataFrame to list-like
:type reduce_method: str or callable
:param remove_empty: if will remove all-zero seqs from output seq_table
:type remove_empty: bool</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.ReactedFractionNormalizer.func">
<em class="property">static </em><code class="sig-name descname">func</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em>, <em class="sig-param"><span class="n">input_samples</span></em>, <em class="sig-param"><span class="n">reduce_method</span><span class="o">=</span><span class="default_value">'median'</span></em>, <em class="sig-param"><span class="n">remove_empty</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.ReactedFractionNormalizer.func" title="Permalink to this definition">¶</a></dt>
<dd><p>core function, input <cite>pd.Dataframe</cite>, output an transformed <cite>pd.Dataframe</cite></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.transform.SpikeInNormalizer">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.transform.</code><code class="sig-name descname">SpikeInNormalizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">spike_in_seq</span></em>, <em class="sig-param"><span class="n">spike_in_amount</span></em>, <em class="sig-param"><span class="n">base_table</span></em>, <em class="sig-param"><span class="n">radius</span></em>, <em class="sig-param"><span class="n">unit</span></em>, <em class="sig-param"><span class="n">dist_type</span><span class="o">=</span><span class="default_value">'edit'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.SpikeInNormalizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.transform.Transformer" title="k_seq.data.transform.Transformer"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.transform.Transformer</span></code></a></p>
<p>Normalized counts to absolute amount using spike-in information</p>
<dl class="py attribute">
<dt id="k_seq.data.transform.SpikeInNormalizer.peak">
<code class="sig-name descname">peak</code><a class="headerlink" href="#k_seq.data.transform.SpikeInNormalizer.peak" title="Permalink to this definition">¶</a></dt>
<dd><p>a Peak instance in <cite>seq-landscape</cite> package to calculate sequence distances</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>landscape.Peak</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="k_seq.data.transform.SpikeInNormalizer.spike_in_amount">
<code class="sig-name descname">spike_in_amount</code><a class="headerlink" href="#k_seq.data.transform.SpikeInNormalizer.spike_in_amount" title="Permalink to this definition">¶</a></dt>
<dd><p>measured spike-in amount in samples</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.Series</p>
</dd>
</dl>
</dd></dl>

<dl class="simple">
<dt>base_table (pd.DataFrame): base_table includes spike-in sequences to calculate normalization factor</dt><dd><p>radius (int): Radius of spike-in peak, seqs less or equal to radius away from center are spike-in seqs
unit (str): unit of spike-in amount
dist_type (str): ‘edit’ or ‘hamming’ as distance measure. Default ‘edit’ to include insertion / deletion</p>
</dd>
</dl>
<dl class="py method">
<dt id="k_seq.data.transform.SpikeInNormalizer.plot_spike_in_peak">
<code class="sig-name descname">plot_spike_in_peak</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.SpikeInNormalizer.plot_spike_in_peak" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the shape of spike-in sequences for each sample</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.SpikeInNormalizer.apply">
<code class="sig-name descname">apply</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.SpikeInNormalizer.apply" title="Permalink to this definition">¶</a></dt>
<dd><p>apply the normalization to a <cite>name</cite> seq_table</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.SpikeInNormalizer._get_mask">
<code class="sig-name descname">_get_mask</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.SpikeInNormalizer._get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>static method to calculate normalization</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.SpikeInNormalizer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">spike_in_seq</span></em>, <em class="sig-param"><span class="n">spike_in_amount</span></em>, <em class="sig-param"><span class="n">base_table</span></em>, <em class="sig-param"><span class="n">radius</span></em>, <em class="sig-param"><span class="n">unit</span></em>, <em class="sig-param"><span class="n">dist_type</span><span class="o">=</span><span class="default_value">'edit'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.SpikeInNormalizer.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize a SpikeInNormalizer</p>
<p>Args:</p>
<dl class="simple">
<dt>spike_in_seq (str): center sequence for spike-in</dt><dd><p>spike_in_amount (list-like, dict, or pd.Series): added spike_in amount, dict and pd.Series should have key of samples in base_table list-like should have same length as number of samples (cols) in base_table
base_table (pd.DataFrame): base_table includes spike-in sequences to calculate normalization factor
radius (int): Radius of spike-in peak, seqs less or equal to radius away from center are spike-in seqs
unit (str): unit of spike-in amount
dist_type (str): ‘edit’ or ‘hamming’ as distance measure. Default ‘edit’ to include insertion / deletion</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="id0">
<code class="sig-name descname">apply</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em><span class="sig-paren">)</span><a class="headerlink" href="#id0" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply normalization to target</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.SpikeInNormalizer.func">
<em class="property">static </em><code class="sig-name descname">func</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em>, <em class="sig-param"><span class="n">norm_factor</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.SpikeInNormalizer.func" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize counts in <cite>name</cite> by <cite>norm_factor</cite></p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.SpikeInNormalizer.radius">
<em class="property">property </em><code class="sig-name descname">radius</code><a class="headerlink" href="#k_seq.data.transform.SpikeInNormalizer.radius" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="id1">
<em class="property">property </em><code class="sig-name descname">spike_in_amount</code><a class="headerlink" href="#id1" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.transform.TotalAmountNormalizer">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.transform.</code><code class="sig-name descname">TotalAmountNormalizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">total_amounts</span></em>, <em class="sig-param"><span class="n">full_table</span></em>, <em class="sig-param"><span class="n">unit</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.TotalAmountNormalizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#k_seq.data.transform.Transformer" title="k_seq.data.transform.Transformer"><code class="xref py py-class docutils literal notranslate"><span class="pre">k_seq.data.transform.Transformer</span></code></a></p>
<p>Quantify the DNA amount by total DNA amount measured in each sample</p>
<p>Amount for each sequence were normalized by direct fraction of the sequence
seq amount = (seq counts / total counts) * total amount</p>
<p>Attributes:
full_table (pd.DataFrame): seq_table where the total amount were measured and normalize to</p>
<blockquote>
<div><p>total_amounts (dict or pd.Series): Total DNA amount for samples measured in experiment
norm_factor (dict): Amount per sequence. Sequence amount = norm_factor * reads
unit (str): Unit of amount measured</p>
</div></blockquote>
<dl class="py method">
<dt id="k_seq.data.transform.TotalAmountNormalizer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">total_amounts</span></em>, <em class="sig-param"><span class="n">full_table</span></em>, <em class="sig-param"><span class="n">unit</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.TotalAmountNormalizer.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize a TotalAmountNormalizer
Args:
total_amounts (dict or pd.Series): Total DNA amount for samples measured in experiment</p>
<blockquote>
<div><p>full_table (pd.DataFrame): seq_table where the total amount were measured and normalize to
name:
unit (str): Unit of amount measured</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.TotalAmountNormalizer.apply">
<code class="sig-name descname">apply</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.TotalAmountNormalizer.apply" title="Permalink to this definition">¶</a></dt>
<dd><p>Transform counts to absolute amount on columns in name that are in norm_factor</p>
<p>Args:
name:</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>pd.DataFrame of normalized name seq_table with only samples in norm_factor</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.TotalAmountNormalizer.full_table">
<em class="property">property </em><code class="sig-name descname">full_table</code><a class="headerlink" href="#k_seq.data.transform.TotalAmountNormalizer.full_table" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.TotalAmountNormalizer.func">
<em class="property">static </em><code class="sig-name descname">func</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em>, <em class="sig-param"><span class="n">norm_factor</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.TotalAmountNormalizer.func" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize name seq_table w.r.t norm_factor</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>pd.DataFrame of normalized name seq_table with only samples provided in norm_factor</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.TotalAmountNormalizer.total_amounts">
<em class="property">property </em><code class="sig-name descname">total_amounts</code><a class="headerlink" href="#k_seq.data.transform.TotalAmountNormalizer.total_amounts" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="k_seq.data.transform.Transformer">
<em class="property">class </em><code class="sig-prename descclassname">k_seq.data.transform.</code><code class="sig-name descname">Transformer</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.Transformer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">abc.ABC</span></code></p>
<p>Abstract class type for transformer</p>
<p>Transformers are classes transform a seq_table instance (<cite>pd.DataFrame</cite> or <cite>SeqTable</cite>) to another seq_table</p>
<dl class="simple">
<dt>To write your transformer, components are:</dt><dd><ul class="simple">
<li><p>Attributes to store parameters</p></li>
<li><p>A static <cite>_get_mask</cite> function to perform transformation</p></li>
<li><p>A <cite>apply</cite> wrapper function for ease of use</p></li>
<li><p>Other Transformer specific utility and visualization functions</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="k_seq.data.transform.Transformer.apply">
<em class="property">abstract </em><code class="sig-name descname">apply</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.Transformer.apply" title="Permalink to this definition">¶</a></dt>
<dd><p>Run the transformation, with class attributes or arguments
Logic and preprocessing of data and arguments should be done here</p>
</dd></dl>

<dl class="py method">
<dt id="k_seq.data.transform.Transformer.func">
<em class="property">abstract static </em><code class="sig-name descname">func</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#k_seq.data.transform.Transformer.func" title="Permalink to this definition">¶</a></dt>
<dd><p>core function, input <cite>pd.Dataframe</cite>, output an transformed <cite>pd.Dataframe</cite></p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="k-seq-data-visualizer-module">
<h2>k_seq.data.visualizer module<a class="headerlink" href="#k-seq-data-visualizer-module" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="k_seq.model.html" class="btn btn-neutral float-right" title="k_seq.model: kinetics and pool models" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="k-seq_package.html" class="btn btn-neutral float-left" title="k-seq package content" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Yuning Shen

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>